docs/content/guides/dagster/local-to-production-tutorial.mdx

```diff
@@ -0,0 +1,448 @@
+        ).json()
+
+
+@resource
+def hn_api_client():
+    return HNAPIClient()
+```
+
+We'll also need to update the `items` asset to use this resource:
+
+```python file=/guides/dagster/local_to_prod_tutorial/assets_v2.py startafter=start_items endbefore=end_items
+# assets.py
+
+
+@asset(
+    config_schema={"N": int},
+    required_resource_keys={"hn_client"},
+)
+def items(context) -> pd.DataFrame:
+    """Items from the Hacker News API: each is a story or a comment on a story."""
+    hn_client = context.resources.hn_client
+
+    max_id = hn_client.fetch_max_item_id()
+    rows = []
+    # Hacker News API is 1-indexed, so adjust range by 1
+    for item_id in range(max_id - context.op_config["N"] + 1, max_id + 1):
+        rows.append(hn_client.fetch_item_by_id(item_id))
+
+    result = pd.DataFrame(rows, columns=hn_client.item_field_names).drop_duplicates(
+        subset=["id"]
+    )
+    result.rename(columns={"by": "user_id"}, inplace=True)
+    return result
+```
+
+<Note>
+  For the sake of brevity in this guide, we've omitted the implementation of the
+  property
+  <code>item_field_names</code> in <code>HNAPIClient</code>. You can find the full
+  implementation of this resource in the <a href="https://github.com/dagster-io/dagster/tree/master/examples/local_to_production">
+    full code example
+  </a> on GitHub.
+</Note>
+
+Now we can write a stubbed version of the Hacker News resource. We want to make sure the stub has implementations for each method `HNAPIClient` implements.
+
+<Note>
+  Determining when it makes sense to stub a resource for a unit test can be a
+  topic of much debate. There are certainly some resources where it would be too
+  complicated to write and maintain a stub. For example, it would be difficult
+  to mock a database like Snowflake with a lightweight database since the SQL
+  syntax may vary. In general, if a resource is relatively simple, writing a
+  stub can be really helpful for unit testing the assets and ops that use the
+  resource.
+</Note>
+
+```python file=/guides/dagster/local_to_prod_tutorial/resources/resources_v2.py startafter=start_mock endbefore=end_mock
+# resources.py
+
+
+class StubHNClient:
+    """
+    Hacker News Client that returns fake data
+    """
+
+    def __init__(self):
+        self.data = {
+            1: {
+                "id": 1,
+                "type": "comment",
+                "title": "the first comment",
+                "by": "user1",
+            },
+            2: {"id": 2, "type": "story", "title": "an awesome story", "by": "user2"},
+        }
+
+    def fetch_item_by_id(self, item_id: int) -> Optional[Dict[str, Any]]:
+        return self.data.get(item_id)
+
+    def fetch_max_item_id(self) -> int:
+        return 2
+
+    @property
+    def item_field_names(self):
+        return ["id", "type", "title", "by"]
+
+
+@resource
+def stub_hn_client():
+    return StubHNClient()
+```
+
+<Note>
```

Comment:
There are four callouts in this section alone - that's a lot.

I like that we're using them, but I think we should be more selective here. We don't want users to get callout fatigue - if everything is important, nothing is important.

Thoughts on this:

- The first (refactoring code to use resources) could be a normal paragraph
- The second (`item_field_names` in `HNAPIClient`) could go either way. I like that it highlights a link to GH, which makes me lean towards keeping it, but it could also be a normal paragraph.
- The third (`Determining when to stub a resource`) should be a normal paragraph. Copy will likely have to be shuffled to accommodate this for flow, but this doesn't need to be a callout.
- The fourth (writing an interface) is a good callout candidate.